From 13b38970f7dc372187ea4eafef65ecc285e77db4 Mon Sep 17 00:00:00 2001
From: Steve Jones <steve@…>
Date: Sat, 1 Feb 2014 16:09:51 +0000
Subject: [PATCH] Set EWMH desktop properties on startup.

Calls ewmh_update_current_desktop on startup to set the _NET Without this
change the properties only get set after the workspaces have been
manipulated. Also exclude hidden workspaces (i.e. those starting with
"__" from the workspace index.

Adds tests for startup and workspace switching.
---
 src/ewmh.c                            |  4 +-
 src/main.c                            |  3 ++
 testcases/t/217-NET_CURRENT_DESKTOP.t | 70 +++++++++++++++++++++++++++++++++++
 3 files changed, 76 insertions(+), 1 deletion(-)
 create mode 100644 testcases/t/217-NET_CURRENT_DESKTOP.t

diff --git a/src/ewmh.c b/src/ewmh.c
index 0298de3..b20d1fd 100644
--- a/src/ewmh.c
+++ b/src/ewmh.c
@@ -27,7 +27,9 @@ void ewmh_update_current_desktop(void) {
     TAILQ_FOREACH(output, &(croot->nodes_head), nodes) {
         Con *ws;
         TAILQ_FOREACH(ws, &(output_get_content(output)->nodes_head), nodes) {
-            if (ws == focused_ws) {
+	    if (STARTS_WITH(ws->name, "__"))
+		continue;
+	    if (ws == focused_ws) {
                 xcb_change_property(conn, XCB_PROP_MODE_REPLACE, root,
                         A__NET_CURRENT_DESKTOP, XCB_ATOM_CARDINAL, 32, 1, &idx);
                 return;
diff --git a/src/main.c b/src/main.c
index ec73955..a616413 100644
--- a/src/main.c
+++ b/src/main.c
@@ -760,6 +760,9 @@ int main(int argc, char *argv[]) {
     /* Set up i3 specific atoms like I3_SOCKET_PATH and I3_CONFIG_PATH */
     x_set_i3_atoms();
     ewmh_update_workarea();
+    
+    /* Set the _NET_CURRENT_DESKTOP property. */
+    ewmh_update_current_desktop();
 
     struct ev_io *xcb_watcher = scalloc(sizeof(struct ev_io));
     struct ev_io *xkb = scalloc(sizeof(struct ev_io));
diff --git a/testcases/t/217-NET_CURRENT_DESKTOP.t b/testcases/t/217-NET_CURRENT_DESKTOP.t
new file mode 100644
index 0000000..a93a2ba
--- /dev/null
+++ b/testcases/t/217-NET_CURRENT_DESKTOP.t
@@ -0,0 +1,70 @@
+#!perl
+# vim:ts=4:sw=4:expandtab
+#
+# Please read the following documents before working on tests:
+# • http://build.i3wm.org/docs/testsuite.html
+#   (or docs/testsuite)
+#
+# • http://build.i3wm.org/docs/lib-i3test.html
+#   (alternatively: perldoc ./testcases/lib/i3test.pm)
+#
+# • http://build.i3wm.org/docs/ipc.html
+#   (or docs/ipc)
+#
+# • http://onyxneon.com/books/modern_perl/modern_perl_a4.pdf
+#   (unless you are already familiar with Perl)
+#
+# Verifies the _NET_CURRENT_DESKTOP property behaves correctly.
+
+use i3test i3_autostart => 0;
+use X11::XCB qw(PROP_MODE_REPLACE);
+use Data::Dumper;
+
+my $config = <<EOT;
+# i3 config file (v4)
+font font -misc-fixed-medium-r-normal--13-120-75-75-C-70-iso10646-1
+fake-outputs 1024x768+0+0
+EOT
+
+my $root = $x->get_root_window;
+my $atom_cookie = $x->intern_atom(0, length("_NET_CURRENT_DESKTOP"), "_NET_CURRENT_DESKTOP");
+my $atom_reply = $x->intern_atom_reply($atom_cookie->{sequence});
+
+my $_NET_CURRENT_DESKTOP = $x->atom(name => "_NET_CURRENT_DESKTOP");
+my $CARDINAL = $x->atom(name => "CARDINAL");
+
+$x->delete_property($root, $_NET_CURRENT_DESKTOP->id);
+
+$x->flush();
+
+sub is_current_desktop_set_to {
+  my $to = shift;
+
+  my $cookie = $x->get_property(0, $root, $x->atom(name => '_NET_CURRENT_DESKTOP')->id,
+				$x->atom(name => 'CARDINAL')->id, 0, 1);
+  my $reply = $x->get_property_reply($cookie->{sequence});
+
+  return 0 if $reply->{value_len} != 1;
+  return 0 if $reply->{format} != 32;
+  return 0 if $reply->{type} != $x->atom(name => 'CARDINAL')->id;
+  my $value = unpack 'L', $reply->{value};
+  return $value == $to;
+}
+
+my $pid = launch_with_config($config);
+
+ok(is_current_desktop_set_to(0), "Starting on desktop 0");
+cmd('workspace 1');
+ok(is_current_desktop_set_to(0), "Change from empty to empty");
+open_window;
+cmd('workspace 0');
+ok(is_current_desktop_set_to(0), "Open on 1 and view 0");
+open_window;
+cmd('workspace 1');
+ok(is_current_desktop_set_to(1), "Open on 0 and view 1");
+cmd('workspace 2');
+ok(is_current_desktop_set_to(2), "Open and view empty");
+
+exit_gracefully($pid);
+
+done_testing;
-- 
1.8.1.2

